opslevel/resource_opslevel_team.go

```diff
@@ -1,311 +1,322 @@
-// 	}
-
-// 	if d.HasChange("parent") {
-// 		if parent := d.Get("parent"); parent != "" {
-// 			input.ParentTeam = opslevel.NewIdentifier(parent.(string))
-// 		} else {
-// 			input.ParentTeam = opslevel.NewIdentifier()
-// 		}
-// 	}
-
-// 	resource, err := client.UpdateTeam(input)
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	if d.HasChange("aliases") {
-// 		err = reconcileTeamAliases(d, resource, client)
-// 		if err != nil {
-// 			return err
-// 		}
-// 	}
-
-// 	if d.HasChange("member") {
-// 		membersErr := reconcileTeamMembership(d, resource, client)
-// 		if membersErr != nil {
-// 			return membersErr
-// 		}
-// 	}
-
-// 	d.Set("last_updated", timeLastUpdated())
-// 	return resourceTeamRead(d, client)
-// }
-
-// func resourceTeamDelete(d *schema.ResourceData, client *opslevel.Client) error {
-// 	id := d.Id()
-// 	err := client.DeleteTeam(id)
-// 	if err != nil {
-// 		return err
-// 	}
-// 	d.SetId("")
-// 	return nil
-// }
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/opslevel/opslevel-go/v2024"
+)
+
+var _ resource.ResourceWithConfigure = &TeamResource{}
+
+var _ resource.ResourceWithImportState = &TeamResource{}
+
+type TeamResource struct {
+	CommonResourceClient
+}
+
+func NewTeamResource() resource.Resource {
+	return &TeamResource{}
+}
+
+// TeamResourceModel describes the Team managed resource.
+type TeamResourceModel struct {
+	Aliases          types.List   `tfsdk:"aliases"`
+	Id               types.String `tfsdk:"id"`
+	LastUpdated      types.String `tfsdk:"last_updated"`
+	Member           []TeamMember `tfsdk:"member"`
+	Name             types.String `tfsdk:"name"`
+	Parent           types.String `tfsdk:"parent"`
+	Responsibilities types.String `tfsdk:"responsibilities"`
+}
+
+type TeamMember struct {
+	Email types.String `tfsdk:"email"`
+	Role  types.String `tfsdk:"role"`
+}
+
+func convertTeamMember(teamMember opslevel.TeamMembership) TeamMember {
+	return TeamMember{
+		Email: RequiredStringValue(teamMember.User.Email),
+		Role:  RequiredStringValue(teamMember.Role),
+	}
+}
+
+func NewTeamResourceModel(ctx context.Context, team opslevel.Team, parent string) (TeamResourceModel, diag.Diagnostics) {
+	aliases, diags := OptionalStringListValue(ctx, team.ManagedAliases)
+	if diags != nil && diags.HasError() {
+		return TeamResourceModel{}, diags
+	}
+	if len(aliases.Elements()) == 0 {
+		aliases = types.ListNull(types.StringType)
+	}
```

Comment:
Is there a case where this if statement is needed? `OptionalStringListValue()` should handle this if `team.ManagedAliases` is empty.
```go
func OptionalStringListValue(ctx context.Context, value []string) (basetypes.ListValue, diag.Diagnostics) {
	if value == nil {
		return types.ListNull(types.StringType), diag.Diagnostics{}
	}
	return types.ListValueFrom(ctx, types.StringType, value)
}
```

Reply:
@davidbloss yes - this is to avoid this error:

```
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to opslevel_team.tf_team, provider "provider[\"registry.terraform.io/opslevel/opslevel\"]" produced an unexpected new value: .aliases: was null, but now cty.ListValEmpty(cty.String).
```

[More context](https://jk-labs.slack.com/archives/C03MX2X6CE5/p1712670162691649?thread_ts=1712610632.527729&cid=C03MX2X6CE5)

[More context 2](https://jk-labs.slack.com/archives/C03MX2X6CE5/p1712684534152499)

In the second thread I proposed that `OptionalStringListValue` should return null instead of empty if `len == 0`.

Reply:
Interesting, we use [OptionalStringListValue() with the infrastructure resource aliases](https://github.com/OpsLevel/terraform-provider-opslevel/blob/feature/migrate-to-tf-framework/opslevel/resource_opslevel_infra.go#L78) and it works fine